### PR TITLE
bulk transition

### PR DESCRIPTION
**Pull request by ar4s**
- [x] working old tests
- [ ] new tests
- [x] refactoring

### CODE REVIEW

File: src/ralph/lib/transitions/admin.py [hunk 1]
```diff
@@ -51,13 +61,47 @@ def change_view(self, request, object_id, form_url='', extra_context=None):
...
+    def get_admin_action_from_transition(self, transition):
+        name = transition.name
+
+        def transition_action_redirect(cls, request, queryset, transition):
+            base_url = reverse(
+                self.get_transition_bulk_url_name(),
+                args=(transition.pk,)
+            )
+            return HttpResponseRedirect(
+                base_url + '?' + '&'.join([
```
Comment (szok) [inline src/ralph/lib/transitions/admin.py]: use: https://docs.python.org/3/library/urllib.parse.html#urllib.parse.urlencode

File: src/ralph/back_office/models.py [hunk 2]
```diff
@@ -274,60 +285,70 @@ def change_user_and_owner(self, **kwargs):
...
                 'now': datetime.datetime.now(),
                 'logged_user': obj_to_dict(request.user),
-                'affected_user': obj_to_dict(self.user),
-                'assets': [{
-                    'sn': self.sn,
-                    'barcode': self.barcode,
-                    'model': str(self.model),
-                    'imei': self.imei or '-'
-                }]
+                # TODO: user validation
```
Comment (mkurek) [inline src/ralph/back_office/models.py]: what about it?

File: src/ralph/lib/transitions/models.py [hunk 3]
```diff
@@ -66,35 +66,93 @@ def run_field_transition(instance, transition, field, data={}, **kwargs):
...
                 if isinstance(field, forms.ChoiceField):
                     value = dict(field.choices).get(int(v))
-
                 field_name = field.label
-            history_kwargs[str(field_name)] = value
+            history[str(field_name)] = value
+    return history
+
+
+def _check_type_instances(instances):
```
Comment (mkurek) [inline src/ralph/lib/transitions/models.py:76]: docstrings would be useful for functions here :)

File: src/ralph/lib/transitions/tests/test_admin.py [hunk 4]
```diff
@@ -19,17 +19,17 @@ def test_url_for_transition(self):
...
-        order = Order.objects.create()
-        transition_url = reverse(
-            'admin:tests_order_transition',
-            args=(order.pk, transition.pk,)
-        )
-        self.login_as_user()
-        self.assertFalse(self.client.session.get('attachment_to_download'))
-        self.client.post(transition_url)
-        self.assertTrue(self.client.session.get('attachment_to_download'))
+    # def test_should_save_attachment_url_in_session(self):
```
Comment (mkurek) [inline src/ralph/lib/transitions/tests/test_admin.py]: ?

File: src/ralph/lib/transitions/templates/transitions/run_bulk_transition.html [hunk 5]
```diff
@@ -0,0 +1,44 @@
+{% extends 'admin/base_site.html' %}
```
Comment (szok) [inline src/ralph/lib/transitions/templates/transitions/run_bulk_transition.html]: why two template? They are the same as the single transition

File: src/ralph/lib/transitions/models.py [hunk 6]
```diff
@@ -16,44 +17,43 @@
...
-    ).choices.from_id(int(transition_field_value)).name
-    setattr(instance, transition_field, int(transition.target))
-    attachment = None
-    action_names = []
-    history_kwargs = {}
-    for action in transition.actions.all():
-        func = getattr(instance, action.name)
-        if not func:
-            continue
+def _save_transition_history(
```
Comment (szok) [inline src/ralph/lib/transitions/models.py]: maybe better bulk add transition history?

File: src/ralph/lib/transitions/models.py [hunk 7]
```diff
@@ -66,35 +67,115 @@ def run_field_transition(instance, transition, field, data={}, **kwargs):
...
+    for action in transition.actions.all():
+        func = getattr(first_instance, action.name)
+        defaults = {
+            key.split('__')[1]: value
+            for key, value in data.items()
+            if key.startswith(action.name)
+        }
         defaults.update(kwargs)
-        result = func(**defaults)
+        # TODO: transaction
```
Comment (mkurek) [inline src/ralph/lib/transitions/models.py:173]: so why don't you use `transaction.atomic` as a decorator on this function?

File: src/ralph/lib/transitions/templates/transitions/run_transition.html [hunk 8]
```diff
@@ -44,4 +49,15 @@
   <script type="text/javascript" src="{% static 'src/js/ralph-autocomplete.js' %}"></script>
   <script type="text/javascript" src="{% static 'vendor/js/foundation-datepicker.js' %}"></script>
   <script type="text/javascript" src="{% static 'src/js/foundation-datepicker-init.js' %}"></script>
+  <script type="text/javascript">
+    (function($) {
```
Comment (mkurek) [inline src/ralph/lib/transitions/templates/transitions/run_transition.html:54]: please add some comment about it (ex. `automatically assign owner to be the same as user`)

File: src/ralph/lib/transitions/templates/transitions/run_transition.html [hunk 9]
```diff
@@ -44,4 +49,15 @@
...
+      var $user_field = $('#id_assign_user__user');
+      var $owner_field = $('#id_assign_owner__owner');
+      if ($user_field.length === 1 && $owner_field.length === 1) {
+        $user_field.change(function() {
+          $owner_field.val($(this).val());
```
Comment (mkurek) [inline src/ralph/lib/transitions/templates/transitions/run_transition.html]: IMO owner should be changed iff it's empty

**Comment by mkurek** [comment]
:ok:

**Comment by szok** [comment]
:ok:

**Pull request merged**